File: 155_min_stack/MinStack.cpp
class MinStack {
    public:
        std::stack<int> stack;
        std::stack<int> min_stack;
        int min;
    
        MinStack() { min = INT_MAX; }
    
        void push(int val) {
            if (val <= min) {
                min_stack.push(min);
                min = val;
            }
    
            stack.push(val);
        }
    
        void pop() {
            int prev = stack.top();
    
            if (prev == min) {
                min = min_stack.top();
                min_stack.pop();
            }
    
            stack.pop();
        }
    
        int top() { return stack.top(); }
    
        int getMin() { return min; }
    };
    
    /**
     * Your MinStack object will be instantiated and called as such:
     * MinStack* obj = new MinStack();
     * obj->push(val);
     * obj->pop();
     * int param_3 = obj->top();
     * int param_4 = obj->getMin();
     */

File: 155_min_stack/MinStack.java
class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;
    private int min;

    public MinStack() {
        this.stack = new Stack<>();
        this.minStack = new Stack<>();
        this.min = Integer.MAX_VALUE;
    }
    
    public void push(int val) {
        this.stack.push(val);
        
        if (val <= this.min) {
            this.minStack.push(this.min);
            this.min = val;
        }
    }
    
    public void pop() {
        int prev = this.stack.pop();

        if (prev == this.min) {
            this.min = this.minStack.pop();
        }
    }
    
    public int top() {
        return this.stack.peek();
    }
    
    public int getMin() {
        return this.min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */